tools/wake-unit/term.cpp

```diff
@@ -384,3 +387,250 @@ TEST(term_dumb_gcc_smoke_test) {
+  if (c <= 0xF) return ' ';
+  if (c == '\033') return ' ';
+  return c;
+}
+
+void append_random_utf8_char(std::string& out, wcl::xoshiro_256& rng) {
+  std::uniform_int_distribution<int> length_dist(1, 4);
+  std::uniform_int_distribution<uint8_t> rand_byte(0, 255);
+  int length = length_dist(rng);
+  uint8_t unset0 = 0x80;
+  uint8_t unset1 = 0x40;
+  uint8_t set1 = 0x80;
+  uint8_t unset2 = 0x20;
+  uint8_t set2 = 0xC0;
+  uint8_t unset3 = 0x10;
+  uint8_t set3 = 0xE0;
+  uint8_t unset4 = 0x8;
+  uint8_t set4 = 0xF0;
+  switch (length) {
+    case 1:
+      // if we get a line feed or a vertical tab, we want
+      // to convert it to a newline so we avoid generating
+      // those here and let another function generate them
+      // instead.
+      out += ascii_mods(fix(rand_byte(rng), 0, unset0));
+      return;
+    case 2:
+      out += fix(rand_byte(rng), set2, unset2);
+      out += fix(rand_byte(rng), set1, unset1);
+      return;
+    case 3:
+      out += fix(rand_byte(rng), set3, unset3);
+      out += fix(rand_byte(rng), set1, unset1);
+      out += fix(rand_byte(rng), set1, unset1);
+      return;
+    case 4:
+      out += fix(rand_byte(rng), set4, unset4);
+      out += fix(rand_byte(rng), set1, unset1);
+      out += fix(rand_byte(rng), set1, unset1);
+      out += fix(rand_byte(rng), set1, unset1);
+      return;
+  }
+}
+
+std::pair<std::string, std::string> random_unicode_string(wcl::xoshiro_256& rng) {
+  std::uniform_int_distribution<int> length_dist(0, 50);
+  int length = length_dist(rng);
+  std::string out;
+  for (int i = 0; i < length; ++i) {
+    append_random_utf8_char(out, rng);
+  }
+
+  // copy output now so that it is only copied once
+  std::string copy = out;
+  return {std::move(out), std::move(copy)};
+}
+
+TEST(term_dumb_fuzz) {
+  std::vector<std::pair<std::string, std::string> (*)(wcl::xoshiro_256 & rng)> funcs = {
+      random_ignored_control_seq,   random_valued_control_seq,
+      random_ignored_command,       random_single_character_command,
+      random_two_character_command, random_three_character_command,
+      random_unicode_string};
+  uint64_t seedV = 0xdeadbeefdeadbeef;
+  std::tuple<uint64_t, uint64_t, uint64_t, uint64_t> seed = {seedV, seedV, seedV, seedV};
+  wcl::xoshiro_256 rng(seed);
+  std::uniform_int_distribution<int> func_picker(0, funcs.size() - 1);
+  std::uniform_int_distribution<int> length_dist(0, 50);
+
+  for (int i = 0; i < 10000; ++i) {
+    int length = length_dist(rng);
+    std::string input, expected;
+    std::string functions;
+    for (int j = 0; j < length; ++j) {
+      int idx = func_picker(rng);
+      functions += std::to_string(idx);
+      auto func = funcs[idx];
+      auto p = func(rng);
+      input += p.first;
+      expected += p.second;
+    }
+
+    // Now we just hope they're equal!
+    CONVERTS_TO_DUMB(input, expected)
+        << "\ninput = " << json_escape(input) << "\nfunctions = " << functions << "\n";
+  }
+}
```

Comment:
This beast of a test attempts to murder TermBufInfo. Unfortunately its kind of hard to generate expected strings for the non-dumb case but given that all the changes in this are fully exercised by the dumb filtering (and that's also what we're most concerned about) we just leave it as is.